### ISSUE TITLE
API to re-download resource

### ISSUE THREAD

**Issue by sl5ne** [greasemonkey/greasemonkey#2331]
It would help script authors if there was an api to download a new copy of a resource even if url is the same. The only workaround I'm aware of is to add fake changes to url in metadata but that requires author manually updating instead of being able to add some logic to script body that would re-download automatically.

For example:

```
// ==UserScript==
// @require        http://www.example.com/example.js
// @resource      resourceName1 resource1.png
// @resource      resourceName2 http://www.example.com/resource2.png
// ==/UserScript==

GM_refresh("http://www.example.com/example.js");
GM_refresh("resource1.png");
GM_refresh("http://www.example.com/resource2.png");
```

**Comment by the8472** [comment]
For frequently changing resources you can just use privileged XHR and set the cache headers appropriately so FF can cache it.

**Comment by janekptacijarabaci** [comment]
See also https://github.com/greasemonkey/greasemonkey/issues/1790

**Comment by sl5ne** [comment]
privileged XHR? Do you mean GM_xmlhttpRequest? If so that isn't as good as @resource. If you mean something else can you please explain?

Also using GM_xmlhttpRequest doesn't help with @require file that changes. 

Script author already has control to get new download by changing url but the method of making fake changes to metadata url is bad workaround. Why not give more direct control through api. Making the fake changes every day and creating new version of script every day is bad but if there was api then author can write code to check once a day.

**Comment by the8472** [comment]
> If so that isn't as good as @resource.

Why not?

**Comment by sl5ne** [comment]
> Why not?

cache-control directives are set by server

stale cached copies aren't updated until expired

**Comment by the8472** [comment]
Adding [cache-busting parameters](https://developer.mozilla.org/en/docs/Web/API/XMLHttpRequest/Using_XMLHttpRequest#Bypassing_the_cache) when you want to bypass the cache is common practice.

In principle a flag could be added to explicitly bypass the cache, but why introduce a new parameter not supported by other scripting engines when a cross-browser solution is already available?

**Comment by sl5ne** [comment]
> Adding cache-busting parameters when you want to bypass the cache is common practice.

> In principle a flag could be added to explicitly bypass the cache, but why introduce a new parameter not supported by other scripting engines when a cross-browser solution is already available?

cache-busting parameters is just adding fake changes to url in metadata

**Comment by the8472** [comment]
that's irrelevant if you're using XHRs

**Comment by sl5ne** [comment]
For @resource yes you could try to replace it with GM_xmlhttpRequest although browser cache purges are out of script author's control.

For @require an API to re-download is still needed.

**Comment by the8472** [comment]
I think *require* only downloading once is somewhat intentional so that there is no switcheroo pulled under the user's nose once a script is installed. Dynamic refresh would amount to execution of remote code in a security context above untrusted web contexts.

This is something the AMO guidelines forbid for example.

If anything I would even advocate the opposite direction, adding hashes for each *require* rule, similar to [SRI-hashes](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity)

**Comment by sl5ne** [comment]
Hacker can already achieve remote code execution in many ways such as eval and xhr. Hacker can already use @require the way it is right now. Little would change by adding a API to re-download.

**Comment by the8472** [comment]
@sl5ne, the point is that `@require` downloads a copy of a file only once. That way a user can review the code and trust it. A remote attacker who has control over the source domain cannot modify it after the fact.

Of course someone can put XHR+eval into a script, but that requires the script itself to be malicious/untrustworthy instead of just the hoster of the dependency to be compromised at a future point in time.

Those are different threat models.

You have to consider that `@require`ed scripts may come from external domains not under the control of the script author.


Verifying that `@require`ed scripts match a hash (like SRI) would mean you only have to trust the script author.
`@require` as it is now means you need to trust the script author and trust the hoster of the dependency at a single point in time
`@require` + a `GM_refresh()` would mean you have to trust the script author and the hoster indefinitely.

In other words, your proposal weakens security because security would depend on trusting more parties. My counter-proposal would strengthen security.
